omnipaxos_storage/src/memory.rs

```diff
@@ -0,0 +1,374 @@
+            db_opts.increase_parallelism(4); // Set the amount threads for rocksDB compaction and flushing
+            db_opts.create_if_missing(true); // Creates an database if its missing in the path
+
+            (c_opts, db_opts)
+        }
+    }
+
+    impl<T, S> Storage<T, S> for PersistentState<T, S>
+    where
+        T: Entry + AsBytes + FromBytes,
+        S: Snapshot<T>,
+    {
+        fn append_entry(&mut self, entry: T) -> u64 {
+            //println!("append entry {:?}", entry);
+            let entry_bytes = AsBytes::as_bytes(&entry);
+            match self.c_log.append_msg(entry_bytes) {
+                Ok(x) => x,
+                Err(_e) => 0,
+            }
+        }
+
+        fn append_entries(&mut self, entries: Vec<T>) -> u64 {
+            //println!("append entries!");
+            let mut buf: MessageBuf = MessageBuf::default();
+            for entry in entries {
+                let _ = buf.push(AsBytes::as_bytes(&entry));
+            }
+            self.c_log.append(&mut buf).unwrap();
+            self.get_log_len()
+        }
+
+        fn append_on_prefix(&mut self, from_idx: u64, entries: Vec<T>) -> u64 {
+            //println!("append on prefix!");
+            let _ = self.c_log.truncate(from_idx);
+            self.append_entries(entries)
+        }
+
+        fn get_entries(&self, from: u64, to: u64) -> Vec<T> {
+            println!("get_entries from: {:?} -> to: {:?} ", from, to);
+            let buffer = self
+                .c_log
+                .read(from, ReadLimit::default())
+                .unwrap_or(MessageBuf::default());
+            let mut entries = vec![];
+            let zero_idx_to = to - from;
+            for (idx, msg) in buffer.iter().enumerate() {
+                if (idx as u64) >= zero_idx_to {
+                    break;
+                }
+                entries.push(FromBytes::read_from(msg.payload()).unwrap());
+            }
+            println!("res from get_entries {:?}", entries);
+            entries
+        }
+
+        fn get_log_len(&self) -> u64 {
+            self.c_log.next_offset()
+            //println!("log length: {:?}", res);
+            //res
+        }
+
+        fn get_suffix(&self, from: u64) -> Vec<T> {
+            //println!("get suffix!");
+            self.get_entries(from, self.get_log_len())
+        }
+
+        fn get_promise(&self) -> Ballot {
+            match self.db.get(b"n_prom") {
+                Ok(Some(mut value)) => {
+                    let prom_bytes: &mut [u8] = &mut value;
+                    FromBytes::read_from(prom_bytes).unwrap()
+                }
+                Ok(None) => Ballot::default(),
+                Err(_e) => Ballot::default(),
+            }
+        }
+
+        fn set_promise(&mut self, n_prom: Ballot) {
+            let prom_bytes = AsBytes::as_bytes(&n_prom);
+            self.db.put(b"n_prom", prom_bytes).unwrap()
```

Comment:
Don't use string as the key in rocksdb. Use `u8` and then define and use constants instead e.g. `const NPROM: u8 = 1` etc for all keys.